File: lib/dop/dop.rb
class Object
	def object?
		is_a?(Hash)
	end

	def array?
		is_a?(Array)
	end
end

class Array

	def get(index)
		return nil if index >= self.size 
		self[index]
	end

	def set(index, value)
		push(nil) while index > size 

		return push(value) if index == size 
		self[index] = value
	end

	def keys 
		(0...size).map{|i|i}
	end

	def next_key 
		self.size
	end  

end

class Hash
	def get(key)
	  self[key.to_sym]  
	end 

	def set(key, value)
	  self[key.to_sym] = value   
	end

	def next_key 
		keys.sort.last.to_s.succ.to_sym
	end  
end

module DOP 
	
	def claves(*camino)
		camino.flatten.map{|k| String === k ? k.to_sym : k }
	end

	def next_key(datos)
		datos.keys.sort.last.to_s.succ     
	end

	def get(data, *camino)
		claves(camino).reduce(data){|d, k| d.get(k) }
	end

	def set(datos, *camino, valor)
		return datos if valor.nil?
		return datos if get(datos, camino) == valor

		clave, *resto = claves(camino)
		 
		datos.clone.tap{|nuevo| nuevo.set(clave, resto.empty? ? valor : set(nuevo.get(clave), resto, valor)) }
	end

	def has(datos, *camino)
		!get(datos, camino).nil?
	end 

	def add(datos, *camino, valor)
		aux = get(datos, camino)
		aux.push(valor)
		set(datos, camino, aux)
	end

	def diff(a, b)
		return diff_o(a, b) if a.object? && b.object?  
		return diff_a(a, b) if a.array?  && b.array?
		return b if a != b
		return nil 
	end

	def diff_a(a, b)
		n = [a.length, b.length].max
		(0...n).map{|i| a[i] == b[i] ? nil : diff(a[i], b[i])}
	end

	def diff_o(a, b)
		e = a.array? ? [] : {}
		return e if a == b 

		keys = (a.keys + b.keys).uniq
		keys.inject(e) do |acc, k|
			d = diff( get(a, k), get(b, k) )
			if d.object? && d.empty? || d.nil?
				acc
			else
				set(acc, [k], d)
			end
		end
	end
	
	def path(datos, salida=[])
	  datos.inject([]) do |acc, (k,v)|
		acc + if v.object? 
			path(v, salida + [k])
		elsif v.array?
			v.map.with_index{|a, i| a && salida + path(a, [k ,i]) }.compact.map(&:first)
		else
			[salida + [k]]
		end
	  end  
	end

	def merge_o(a, b)
		b.keys.inject(a){|d,k| set(d, k, merge(a.get(k), b.get(k)))}
	end
	
	def merge_a(a, b)
		b.keys.inject(a){|d,k| set(d, k, merge(a.get(k), b.get(k)))}
	end

	def merge(a, b)
		return a if a == b 
		if a.object? && b.object? || a.array?  && b.array?
			b.keys.inject(a){|d, k| set(d, k, merge(a.get(k), b.get(k)))}	
		else 
			b
		end 
	end

	def igual(a, b)
		if a == b
			true
		elsif a.object? && b.object? || a.array? && b.array?
			a.keys.all? {|k| igual(a.get(k), b.get(k)) }
		end
	end

end



# -----------------------
include DOP 


# probar_get_set(datos)
# Clases

#   Usuario
#   - id
#   - dni
#   - nombre

#   Negocio
#   - id
#   - cuit
#   - denominacion

#   Cuenta
#   - id
#   - usuario
#   - creacion
   
#   Movimiento
#   - origen
#   - destino
#   - monto 

# {:b=>22, :d=>[nil, {:n=>4}, {:o=>3}], :e=>{:y=>{:v=>12, :z=>30}}, :c=>5}
# [[:b], [:d], [:e, :y, :v], [:e, :y, :z], [:c]]
# 
# [[:b], {:n=>4}, {:o=>3}, [:e, :y, :v], [:e, :y, :z], [:c]]
# [[:b], [[:d, 1, :n]], [[:d, 2, :o]], [:e, :y, :v], [:e, :y, :z], [:c]]

# [[:b], [:d, 1, :n], [:d, 2, :o], [:e, :y, :v], [:e, :y, :z], [:c]]
# 
#
#
def probar_diferencia
	a = {a: 10, b: 20,      d:[{m: 1}, {m: 2, n: 3}],         e: {x: 10, y: {v: 10, w: 20}}}
	b = {a: 10, b: 22, c:5, d:[{m: 1}, {m: 2, n: 4}, {o: 3}], e: {x: 10, y: {v: 12, z: 30}}}
	puts 'Diferencia >> '
	pp d = diff(a, b)
	# --
	# pp [:b, :d, [[:e, :y, :v],[:e, :y, :z]], :c]
	# --

	puts "--------"
	puts " > A "
	pp a 
	puts " > B "
	pp b 
	puts " A <=> B"
	pp d 
	puts "--"
	pp path(d)

	puts "Merge"

	pp merge(a, d)
end


probar_diferencia

File: lib/dop/dop.dart
typedef MapDyn = Map<String, dynamic>;
typedef ListDyn = List<dynamic>;

extension ObjectExtension on Object {
  bool get isObject => this is Map;
  bool get isArray => this is List;
}

extension ArrayExtension on List {
  dynamic getAt(int index) {
    if (index >= length) {
      return null;
    }
    return this[index];
  }

  void setAt(int index, value) {
    while (index >= length) {
      add(null);
    }
    this[index] = value;
  }

  List<int> get keys => List<int>.generate(length, (i) => i);

  int get nextKey => length;
}

extension MapExtension on Map {
  dynamic get(String key) {
    return this[key];
  }

  void set(String key, value) {
    this[key] = value;
  }

  String nextKey() {
    final sortedKeys = keys.map((key) => key.toString()).toList()..sort();
    if (sortedKeys.isEmpty) {
      return '0';
    }
    return (int.parse(sortedKeys.last) + 1).toString();
  }
}

mixin DOP {
  ListDyn claves(ListDyn camino) {
    return camino.map((k) => k is String ? k.toString() : k).toList();
  }

  String nextKey(MapDyn datos) {
    final sortedKeys = datos.keys.map((key) => key.toString()).toList()..sort();
    if (sortedKeys.isEmpty) {
      return '0';
    }
    return (int.parse(sortedKeys.last) + 1).toString();
  }

  dynamic get(MapDyn data, ListDyn camino) {
    return claves(camino).fold(data, (d, k) => d[k]);
  }

  MapDyn set(MapDyn datos, ListDyn camino, dynamic valor) {
    if (valor == null) {
      return datos;
    }
    if (get(datos, camino) == valor) {
      return datos;
    }

    final clave = claves(camino).first;
    final resto = claves(camino).sublist(1);
    final nuevo = Map.from(datos) as MapDyn;
    nuevo[clave] = resto.isEmpty ? valor : set(nuevo[clave] ?? {}, resto, valor);
    return nuevo;
  }

  bool has(MapDyn datos, ListDyn camino) {
    return get(datos, camino) != null;
  }

  MapDyn add(MapDyn datos, ListDyn camino, dynamic valor) {
    final aux = get(datos, camino) ?? <dynamic>[];
    aux.add(valor);
    return set(datos, camino, aux);
  }

  dynamic diff(dynamic a, dynamic b) {
    if (a is MapDyn && b is MapDyn) {
      return diffObject(a, b);
    } else if (a is ListDyn && b is ListDyn) {
      return diffArray(a, b);
    } else if (a != b) {
      return b;
    } else {
      return null;
    }
  }

  ListDyn diffArray(ListDyn a, ListDyn b) {
    final n = a.length > b.length ? a.length : b.length;
    final diffList = <dynamic>[];
    for (int i = 0; i < n; i++) {
      diffList.add(a[i] == b[i] ? null : diff(a[i], b[i]));
    }
    return diffList;
  }

  MapDyn diffObject(MapDyn a, MapDyn b) {
    final e = a.isEmpty && b.isEmpty ? {} as MapDyn : a;
    if (a != b) {
      final keys = a.keys.toSet()..addAll(b.keys);
      for (final key in keys) {
        final d = diff(a[key], b[key]);
        if (d is Map && d.isEmpty || d == null) {
          continue;
        }
        e[key] = d;
      }
    }
    return e;
  }

  List<ListDyn> path(MapDyn datos, ListDyn salida) {
    final result = <ListDyn>[];
    for (final entry in datos.entries) {
      final k = entry.key;
      final v = entry.value;
      if (v is MapDyn) {
        result.addAll(path(v, [...salida, k]));
      } else if (v is List) {
        for (int i = 0; i < v.length; i++) {
          if (v[i] != null) {
            result.add([...salida, k, i]);
          }
        }
      } else {
        result.add([...salida, k]);
      }
    }
    return result;
  }

  MapDyn mergeObject(MapDyn a, MapDyn b) {
    b.forEach((key, value) {
      a[key] = merge(a[key] ?? {}, value);
    });
    return a;
  }

  ListDyn mergeArray(ListDyn a, ListDyn b) {
    for (int i = 0; i < b.length; i++) {
      a[i] = merge(a[i] ?? {}, b[i]);
    }
    return a;
  }

  dynamic merge(dynamic a, dynamic b) {
    if (a == b) {
      return a;
    } else if ((a is MapDyn && b is MapDyn) || (a is ListDyn && b is ListDyn)) {
      if (a is! Map) {
        a = <String, dynamic>{};
      }
      if (b is! Map) {
        b = <String, dynamic>{};
      }
      if (a is! List) {
        a = <dynamic>[];
      }
      if (b is! ListDyn) {
        b = <dynamic>[];
      }
      if (a is MapDyn) {
        return mergeObject(a, b);
      } else {
        return mergeArray(a, b);
      }
    } else {
      return b;
    }
  }

  bool igual(dynamic a, dynamic b) {
    if (a == b) {
      return true;
    } else if (a is MapDyn && b is MapDyn) {
      return a.keys.every((key) => igual(a[key], b[key]));
    }
    return false;
  }
}
